### ISSUE TITLE
Proposal for a Simplified I18N Process

### ISSUE THREAD

**Issue by HansMuller** [flutter/flutter#41437]
## Introduction
This is a proposal to provide a tool that simplifies setting up and maintaining an internationalized Flutter app. It is based on the Dart intl package. The overall goal is to reduce the number of steps required to use that package in an app with modest i18n requirements.

## Current Internationalization Process
Currently, internationalized apps which are based on the intl package must follow a procedure similar to the one outlined in the [stocks example](https://github.com/flutter/flutter/blob/master/examples/stocks/lib/i18n/regenerate.md). Roughly speaking there are 5 steps:

- Add calls to the static `Intl.message` and `Intl.plural` methods where message strings are needed by the app.
- Generate a template ".arb" (application resource bundle) file with `intl_translation:extract_to_arb`.
- Make copies of the template ".arb" file for each supported locale, and add message translations.
- Generate several source files which collectively encode all of the localizations using `intl_translation:generate_from_arb`.
- Before the app starts, call `initializeMessages()` for each supported locale.

## Proposal
The proposed approach is to start with handwritten ".arb" files. One of these files will contain the standard per-message metadata required by the intl package.

A new tool will be used to generate a single class that contains one method per message.

Message lookup is based on the inherited locale, using the current BuildContext:

```dart
MyLocalizations.of(context).myMessage()
```

The generated class will provide a LocalizationDelegate which is to be included in the app's list of localization delegates. The generated class itself will depend on intl package; the rest of the app doesn't have to.

This approach, based on segregating the app's localizations in a single class, is similar to the stocks example.

## Prototype App and Tool
A prototype of the tool is https://github.com/flutter/flutter/pull/41432.

The main elements of the app, including the generated code, are as follows.

The English locale ".arb" file, [lib/l10n/demo_en.arb](https://gist.github.com/HansMuller/064e6e3633d4dc608ab0baa576fd6e73) defines the messages. The Spanish locale arb file, [lib/l10n/demo_es.arb](https://gist.github.com/HansMuller/4beb7f05c99b49398a4812200b62d411), just defines message translations.

The generated localizations class, DemoLocalizations in [lib/1l0n/demo_localizations.dart](https://gist.github.com/HansMuller/924ec00e80bff32ba8b983e88b587d81), provides all of the messages.

The [app](https://gist.github.com/HansMuller/883d6f9d6a69a094f67d5bf9c93dd300) includes `DemoLocalizations.delegate` in its `localizationDelegates` list, and looks up its messages with `DemoLocalizations.of(context)`.

**Comment by escamoteur** [comment]
Have you seen this package https://github.com/timfreiheit/r_flutter?
Also just a thought wouldn't it even more comfortable if the input would be one cvs file with all languages in it that easy can be edited with any spread sheet application, so easy to give to translators.
Also have you seen https://marketplace.visualstudio.com/items?itemName=esskar.vscode-flutter-i18n-json

To make is really comfortable you should be able to directly create a new word entry from the IDE.

**Comment by alan-knight** [comment]
The target audience for this are people who doing their own translations by directly modifying the data files? I would have thought most people were using some sort of tool or translation service, but it seems not. Given that it seems reasonable enough. It seems important to consider migration issues when someone scales up to using a tool/service. I think they could probably still generate that way, but could this reasonably support multiple classes/multiple ARB files? Is that something people need/use in larger Flutter applications?

Intl doesn't currently handle formatting of parameters, so if you have something that takes a date or a number, you either have to format it before passing it in as a parameter or it will get the default format. So the rest of the app might still need to depend on intl for that.

Not directly relevant to this proposal, but Google Translator Toolkit, which is the primary thing that supported the ARB format is deprecated. We will really need to support at least one other format. There is a pending bug on intl_translation to make it so that anyone can write a reader/writer external to the package. There's also a converter someone wrote in JS, between ARB files and "Hierarchical JSON" format and it could probably handle others.

CSV files with Unicode in them makes me nervous of escaping issues, but there are a lot of formats. As above, being able to handle lots of them seems good, but we may want to recommend a default.

Does this/can this reasonably combine with the ability to use native translations on mobile platforms? So, would people be more comfortable with editing e.g. Android resource files directly?

There's been a bit of work ongoing on being able to read ICU MessageFormat strings at runtime and format them, as Closure does. That would be another alternative, although it's an entirely different (and much less typed) API so would be a significant change. And those would still have to be embedded in a file format.

**Comment by escamoteur** [comment]
@alan-knight what is the problem with CVS and Unicode? 
I still think a format that can be read by office applications would make it much easier.

Another point not to forget is how can we support singular / plural versions of words in different languages?

**Comment by HansMuller** [comment]
> It seems important to consider migration issues when someone scales up to using a tool/service. I think they could probably still generate that way, but could this reasonably support multiple classes/multiple ARB files? Is that something people need/use in larger Flutter applications?

The tool could definitely support generating multiple classes for different sets of ARB files.  That does seems like a nice kind of flexibility. I don't know if developers need that kind of support now, although I'm hoping we'll hear from them here.


> Intl doesn't currently handle formatting of parameters, so if you have something that takes a date or a number, you either have to format it before passing it in as a parameter or it will get the default format. So the rest of the app might still need to depend on intl for that.

Yes, that's correct.

> Not directly relevant to this proposal, but Google Translator Toolkit, which is the primary thing that supported the ARB format is deprecated. We will really need to support at least one other format. There is a pending bug on intl_translation to make it so that anyone can write a reader/writer external to the package. There's also a converter someone wrote in JS, between ARB files and "Hierarchical JSON" format and it could probably handle others.
> ...
> Does this/can this reasonably combine with the ability to use native translations on mobile platforms? So, would people be more comfortable with editing e.g. Android resource files directly?

The ARB format isn't essential to this proposal, however it's what we've recommended that Flutter developers use to date. We could support native message catalog formats, per https://github.com/flutter/flutter/issues/37031, as well.

> There's been a bit of work ongoing on being able to read ICU MessageFormat strings at runtime and format them, as Closure does. That would be another alternative, although it's an entirely different (and much less typed) API so would be a significant change. And those would still have to be embedded in a file format.

It would be nice to support a platform-independent and widely accepted standard. Assuming that the right API to expose to an app was still a set of methods that returned strings, supporting ICU (or ICU too) sounds like a win.

**Comment by Tomucha** [comment]
Hi! I'm all for handwritten message files and code generation. The compile-time check of the message existence is great. But don't forget that occasionally you need to decide the message at runtime, maybe like this:

`var vehicleTypeMessageKey = "VehicleType.${data['type']'}";
`
... so having just generated methods is not enough.

You will find a few more ideas like this in my internationalization package :-)

https://pub.dev/packages/i69n

**Comment by ashour** [comment]
Hello, I hope you're all well. Thanks for your work on Flutter btw. It's an awesome piece of tech and I wish I had Flutter when I was co-founding a mobile app startup years ago. I was recently emailed by @HansMuller, who asked me to chime in. I research and write about i18n. I wrote [an article on Flutter + Intl](https://phrase.com/blog/posts/how-to-internationalize-a-flutter-app/) last year (I think that's how Hans found me).

I think this is a good proposal because the current Intl workflow is a bit clunky for smaller projects. This is especially true for projects that don't have automation and sophisticated tooling around i18n. Given this leaner alternative, I think two main pain points from a developer  perspective warrant a look:

- Reducing manual script running as much as possible
- Using quick, lean file formats for messages that can be easily shared with translators

RE reducing manual script running, I've always wondered while using Flutter + Intl if there could be a watcher that generated and updated message files on the fly while developing. I've actually never seen this in any i18n library and I've always wanted it. I don't know how feasible this is given available Dart code analysis tools VS what would need to be written from scratch for such a watcher. 

As for file formats, there are a few I think worth considering (not exclusively, of course, just ones that come to my mind):

- ICU
- PO
- CSV
- JSON
- YAML

ICU and PO are widely used. PO has good adoption among i18n tools and platforms for developers & translators. For a smaller team, however, I think the overhead of generating and compiling ICU or PO is more of a hassle than its worth. The other options are much more lean and to-the-point for smaller projects. 

CSV is attractive for sharing with translators, although I can see it being a pain for developers. CSV headers can be far removed from their respective values, which can make managing messages tricky. Think of a plural message in Arabic with the language's 6 plural forms. How could a developer tell them apart? You could add an identifier to each message component, but that would take away from translators' experience.

JSON would be nice for us coders, but it may be a bit of a hassle for translators.

YAML could be a nice in-betweener for both programmers and translators. YAML has a nice, intuitive syntax with significant white space that I think would be picked up easily by translators. It also allows for various types and hierarchies, which make it flexible enough accommodate plural forms and more complex messages.

It could be worthwhile to pull in some translators that work in software and get their thoughts. They may well have their own bread-and-butter tools that, say, work well with JSON and have wide adoption. In that case JSON would be a very attractive option since we'd really only be solving the developer's problem, not the translators'.

These are my two cents so far. I hope they help a bit.

**Comment by HansMuller** [comment]
@tomucha - Thanks for the feedback. I think we've taken a similar approach to simplifying building i69n applications :-).

> But don't forget that occasionally you need to decide the message at runtime, maybe like this:
> 
> var vehicleTypeMessageKey = "VehicleType.${data['type']'}";
> ... so having just generated methods is not enough.
> 

I agree that this sort of thing isn't necessary to solve with generated methods. I would think the app would have to include a method that switched on `vehicleTypeMessageKey` and returned a localized string.
```dart
String vehicleType(String vehicleTypeMessageKey) {
  final MyLocalizations localizations = MyLocalizations.of(context);
  switch (vehicleTypeMessageKey) {
    case 'sedan': return localizations.sedan();
    case 'truck': return localizations.truck();
    ...
  }
}
```

**Comment by Tomucha** [comment]
At this point i69n generates the exact switch statement you are suggesting ;-)

**Comment by szotp** [comment]
@ashour what's wrong with .arb files? Flutter already uses them, and they are esentially jsons.

@HansMuller I agree that current state is a mess. Did you consider doing `flutter generate` builder package? I implemented this in `r_flutter` and it worked decently well (although IDE & analyzer are not really prepared to handle it). Such package could be eventually embraced by flutter team.

**Comment by HansMuller** [comment]
@ashour - Your [article on internationalizing Flutter apps](https://phrase.com/blog/posts/how-to-internationalize-a-flutter-app/) is an excellent introduction to this topic!

> I think two main pain points from a developer  perspective warrant a look:
> 
> - Reducing manual script running as much as possible
> - Using quick, lean file formats for messages that can be easily shared with translators
> 
> RE reducing manual script running, I've always wondered while using Flutter + Intl if there could be a watcher that generated and updated message files on the fly while developing. I've actually never seen this in any i18n library and I've always wanted it. I don't know how feasible this is given available Dart code analysis tools VS what would need to be written from scratch for such a watcher.

That's an excellent point. We can definitely do this, even in the context-of hot-reload, in roughly the same way that Flutter handles asset changes.

> As for file formats, there are a few I think worth considering (not exclusively, of course, just ones that come to my mind):
> 
> - ICU
> - PO
> - CSV
> - JSON
> - YAML
> 

Currently the ["ARB" format](https://github.com/google/app-resource-bundle/wiki/ApplicationResourceBundleSpecification), which is used by the Dart intl package and the proposed tool, is mostly just a JSON container for a collection of ICU messages. Additional support (beyond what's in the prototype) for dates, currencies, numbers, gender would move us from the realm of "modest" i18n Flutter apps, to more ambitious ones. I think that would be a good direction to take in the long run.

Support for other message catalog formats [came up earlier](https://github.com/flutter/flutter/issues/41437#issuecomment-536114778). We could support other formats with additional code generators or perhaps with a scheme that converts other formats to a common one. [XLIFF](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LocalizingYourApp/LocalizingYourApp.html) is another popular alternative.

> These are my two cents so far. I hope they help a bit.

Thanks for your feedback!

**Comment by HansMuller** [comment]
@szotp - thanks for your feedback.

> I agree that current state is a mess. Did you consider doing flutter generate builder package? I implemented this in r_flutter and it worked decently well (although IDE & analyzer are not really prepared to handle it). Such package could be eventually embraced by flutter team.

I'll take a look at [r_flutter](https://pub.dev/packages/r_flutter).

**Comment by escamoteur** [comment]
Looking back when I did.Net programming there was a powerful free Application to manage the resource files including text resources in different languages. That could also in and export different formats. 

So the used format might not be so important for users if someone writes a flutter desktop app for this. 

So maybe

**Comment by alan-knight** [comment]
Dart has a lot of things that rely on code generation and mechanisms for automatically re-generating on changes, e.g. build_runner. So I would expect automating that to be fairly easy. But I don't think anything is integrated with flutter right now, so I guess you would have to manually run build_runner on the side and have it watch the files.

**Comment by jonahwilliams** [comment]
I don't think we want to require a dependency on build_runner to generate translations.

**Comment by alan-knight** [comment]
No, build_runner is pretty heavy for that. But I don't think it would be difficult to do something custom that did file watching and triggered a re-generate, or to use other mechanisms.

**Comment by escamoteur** [comment]
Just one thing that came to my mind. If this gets sort of service think about adding an interface that can be easily used by the IDE extensions for VS code and IntelliJ.
With a function in the editor to insert a new identifier together with a default language that then creates an entry in the ARB(or whatever) files would make you really productive with internationalisation.

**Comment by samuelg123** [comment]
Hi guys,
I want to share this with you https://github.com/long1eu/flutter_i18n . Works great in Android Studio. I've been using that plugin for about 1 month. Just like the old days when developing Android (especially for alt+enter feature to extract string) :)

**Comment by bitsydarel** [comment]
Simplified I18N Process

Can this task also take care of updating the info.plist about the supported languages so that we won’t have to do it manually for iOS?

**Comment by HansMuller** [comment]
@bitsydarel That's a good idea; yes (https://github.com/flutter/flutter/issues/43041#issuecomment-544565791)

**Comment by bitsydarel** [comment]
Reading the proposal i have few questions and proposal.

1- Why not sticking with arb and just make the files generation and intl initialization process almost inexistant?

Arb it’s used in many translation tools and it’s actually a popular format with many features.

Using csv, json, xml we will have to specify a way of defining plurals, gender and ect.

2- keep the way users specify the properties to be translated  (custom Localization class with properties forwarding to intl.message).

3- automate the creation of arb file template and the one for each languages that’s defined in the supportedLanguages property of the widgetsApp.

In other languages arb files, let the translation empty so users can fill it.

4- generate the binding code to initialize intl.

5- generate the delegates that will load our localization class and initialize intl.

this allow us as developer to only create the localization class and the properties and all the rest is generated for us and no ide required.

Why not using build_runner ?

I can’t think of a real production app that does not use code generation... most of the libraries we daily use, use them, json code generation, SQLite and more.

Using build_runner watch, would allow us to interactively update the arb files while we are developing our app, which is amazing for programmer experience and better than native solutions.

**Comment by hpoul** [comment]
I personally would vote for native support for PO files (generating to/from dart code), but I guess it would be quite trivial to have a json/po converter. I've got used to PO files from python/django and it seems it has a pretty wide range of tools available (e.g. poedit, transifex, etc.).

In my dream setup the source of truth for string keys would be the dart code.  Similar to right now. So during development there is no need to update PO files immediately. Once code is done, extract to PO files, translate the PO files and then generate resulting dart code from those po files.
Imo even in trivial apps it would be very much helpful to allow easy separation into multiple classes. Having all app strings in one file/class can get pretty crowded pretty soon.

**Comment by bitsydarel** [comment]
You can already use multiple localizations classes @hpoul 

The framework only require you to provide delegates for them.

```dart
class DBAssistantLocalizations {
  /// List of locale supported by [DBAssistantLocalizations].
  static const List<Locale> supportedLanguages = <Locale>[
    Locale.fromSubtags(languageCode: "en"),
    Locale.fromSubtags(languageCode: "ru"),
    Locale.fromSubtags(languageCode: "uk")
  ];

  /// Get an instance of [DBAssistantLocalizations] from the current [context].
  static DBAssistantLocalizations of(final BuildContext context) =>
      Localizations.of<DBAssistantLocalizations>(
        context,
        DBAssistantLocalizations,
      );

  // ignore: public_member_api_docs
  String get appTitle => Intl.message(
        "DBAssistant",
        name: "app_title",
        desc: "the application title",
      );

  // ignore: public_member_api_docs
  String get profile => Intl.message(
        "Profile",
        name: "profile",
        desc: "Represent user profile",
      );

  // ignore: public_member_api_docs
  String get high => Intl.message(
        "High",
        name: "high",
        desc: "adjective especially of things that are not living",
      );

  // ignore: public_member_api_docs
  String get medium => Intl.message(
        "Medium",
        name: "medium",
        desc: "being in the middle between an upper"
            " and lower amount, size, degree, or value",
      );

  // ignore: public_member_api_docs
  String get low => Intl.message(
        "Low",
        name: "low",
        desc: "adjective below the usual level",
      );

  // ignore: public_member_api_docs
  String get lists => Intl.message(
        "Lists",
        name: "lists",
        desc: "feature of the app",
      );

  // ignore: public_member_api_docs
  String get news => Intl.message(
        "News",
        name: "news",
        desc: "feature of the app",
      );

  // ignore: public_member_api_docs
  String get tv => Intl.message(
        "Tv",
        name: "tv",
        desc: "feature of the app",
      );

  // ignore: public_member_api_docs
  String get couldNotRetrieveWeatherInfo => Intl.message(
        "Could not retrieve weather information",
        name: "couldNotRetrieveWeatherInfo",
        desc: "Message that for some reason weather date could not be retrieve",
      );

  // ignore: public_member_api_docs
  String get noShoppingLists => Intl.message(
        "You currently don't have any shopping lists",
        name: "no_shopping_lists",
        desc: "The message for when the user does not have any shopping lists",
      );
}
```

```dart
class DBAssistantDateTimeLocalizations {

  /// Get an instance of [DBAssistantDateTimeLocalizations]
  /// from the current [context].
  static DBAssistantDateTimeLocalizations of(final BuildContext context) {
    return Localizations.of<DBAssistantDateTimeLocalizations>(
      context,
      DBAssistantDateTimeLocalizations,
    );
  }

  // ignore: public_member_api_docs
  String years(int years) => Intl.plural(
        years,
        one: "$years year",
        other: "$years years",
        name: "formatted_years",
        desc: "year formatter base on the amount of years",
      );

  // ignore: public_member_api_docs
  String months(int months) => Intl.plural(
        months,
        one: "$months month",
        other: "$months months",
        name: "formatted_months",
        desc: "month formatter base on the amount of months",
      );

  // ignore: public_member_api_docs
  String days(int days) => Intl.plural(
        days,
        one: "$days day",
        other: "$days days",
        name: "formatted_days",
        desc: "day formatter base on the amount of days",
      );

  // ignore: public_member_api_docs
  String hours(int hours) => Intl.plural(
        hours,
        one: "$hours hour",
        other: "$hours hours",
        name: "formatted_hours",
        desc: "hour formatter based on the amount of hours",
      );

  // ignore: public_member_api_docs
  String minutes(int minutes) => Intl.plural(
        minutes,
        one: "$minutes minute",
        other: "$minutes minutes",
        name: "formatted_minutes",
        desc: "minute formatter based on the amount of minutes",
      );

  // ignore: public_member_api_docs
  String get fewMoments => Intl.message(
        "few moments",
        name: "few_moments",
        desc: "refer the time elapsed since an event",
      );

  // ignore: public_member_api_docs
  String getLastChanges(final String elapsedTime) => Intl.message(
        "Last changes: $elapsedTime ago",
        name: "last_changes_formatted",
        desc: "Display the last time a change occured on a object",
        args: [elapsedTime],
        examples: const {"elapsedTime": "2h"},
      );
}
```

**Comment by bitsydarel** [comment]
And ARB has many features that PO does not support.

**Comment by hpoul** [comment]
@bitsydarel 
> You can already use multiple localizations classes

I wasn't disputing that, just wanted to raise that issue because the current proposal mentions:

> A new tool will be used to generate a single class that contains one method per message.

not sure how much emphasis was on that "single class"..

**Comment by HansMuller** [comment]
@bitsydarel per https://github.com/flutter/flutter/issues/41437#issuecomment-544647725

> 1- Why not sticking with arb and just make the files generation and intl initialization process almost inexistant?

Although, in the fullness of time, we hope to provide support for other formats, initially we'll support the arb format.

The "Proposal" section of this issue begins with "The proposed approach is to start with handwritten ".arb" files."

> 2- Keep the way users specify the properties to be translated (Custom Localization class with properties forwarding to intl.message).

That is similar to what's being proposed here. However in this case the custom localization class is generated rather than hand written. See for example, the [code generated by the prototype](https://gist.github.com/HansMuller/924ec00e80bff32ba8b983e88b587d81)  is based on the [sample .arb](https://gist.github.com/HansMuller/064e6e3633d4dc608ab0baa576fd6e73) file.

> 3- Automate the creation of arb file template and the one for each languages that’s defined in the supportedLanguages property of the widgetsApp.

In this proposal the template arb file is generated by hand. With some IDE integration, we could simplify creating it. Simiarly, we could generate copies of the template arb file, per supported locale.

> 4- generate the binding code to initialize intl.
> 5- generate the delegates that will load our localization class and initialize intl.

We are in fact proposing to do this and to do so automatically, when the .arb files change.

I think the main difference between what you're suggesting and what has been proposed is the starting point. We're proposing to start with .arb files, since they're the source of localizations, and generate the rest.

**Comment by HansMuller** [comment]
@hpoul per https://github.com/flutter/flutter/issues/41437#issuecomment-544653694

> I personally would vote for native support for PO files (generating to/from dart code)...

In the fullness of time, we hope to provide support for other formats, initially we'll support the arb format.

> Imo even in trivial apps it would be very much helpful to allow easy separation into multiple classes. Having all app strings in one file/class can get pretty crowded pretty soon.


Making it possble to generate more than one localization class sounds like a good idea.

**Comment by bitsydarel** [comment]
Ho so then i did not fully understand the proposal and I’m happy all the other suggestions are already covered by the initial proposal.

About the starting point, If we have to start by creating an initial arb file this require knowledge of the arb specifications and format.

Which is definetly overwhelming for beginners, people who would like to quickly integrate localization for their apps and someone coming from another platform.

That's why i suggested that it's should be generated based on the content of the custom localization class because this is just dart code that the programmer is already familiar with.

So me as a programmer i just have to worry about writing my custom localization class.

Now for the copies of the template arb for each supported locale, since they would already be generated and pre-filled.
I would just have to fill the translated value i have gotten from a trusted source example: Google translate, translations team, ect.

**Comment by bitsydarel** [comment]
@HansMuller If in the futur you plan on adding support for other formats, the suggested flow would require less work on the developer side because the only changes would be the type of files generated and how the IDE plugin or generator will handle the binding.

**Comment by bitsydarel** [comment]
@HansMuller Now that i think of it, the only issue i see with my suggestion is if let say we add support for PO files or any other format, if the developer wrote the custom localization class property using Intl.message, Intl,plural, ect. Can we still do the binding (Intl initialization) properly ? Or is Intl depending on ARB ?

If Intl is depending on ARB then your proposal is better because with your approach it's does not matter which file format we use, at the end we only need that the custom localization class is properly generated and with proper bindings or initialization so that we could do:

```dart
MyLocalizations.of(context).myMessage()
```

**Comment by marcglasberg** [comment]
This issue was shown to me by @escamoteur. 

I'd like to share with you my approach to translations. I just published it yesterday. It uses the new Dart extensions: https://pub.dev/packages/i18n_extension

Suppose you have this: `Text("Hello")` and you want to translate it to the current locale. You just need to add `.i18n` to the string, like this: `Text("Hello".i18n)`.

This would display "Hola" is the locale is Spanish, "Olá" if it's Portuguese, and so on. You can also have plurals, using the `.number()` modifier. For example `Text("There is %d dog".number(2))` will display "There are 2 dogs". And you can create your own custom modifiers, for things like genders etc.

It's specially easy to use for small projects, where you are doing your own translations.

But it also works well for large projects (like the one I'm involved in, which will use this).

In more detail, you have a dart file that creates a Translation object, and this object is optimized for easily creating translations by hand. But it creates them from maps. So if you can create maps from some file you can use that file. In other words, it's very easy to import AND export from/to `.json,`, `.arb` etc.

Also, there is no need for build_value.

**Comment by HansMuller** [comment]
@bitsydarel per https://github.com/flutter/flutter/issues/41437#issuecomment-544744428

> ... the only issue i see with my suggestion is if let say we add support for PO files or any other format, if the developer wrote the custom localization class property using Intl.message, Intl,plural, ect. Can we still do the binding (Intl initialization) properly ? Or is Intl depending on ARB ?

The Dart intl package depends on the ARB format. That's why this proposal is based on ARB, at least by default. We could support other formats, possibly by converting them to ARB. In that case ARB would just be an intermediate format.

> If Intl is depending on ARB then the current proposal is better because the current proposal it does not matter which file format we use, at the end we only need that the custom localization class is properly generated and with proper bindings or initialization so that we could do:
>
> MyLocalizations.of(context).myMessage()

It's true that we could support other formats directly, by generating a MyLocalizations class that did the work for plurals, genders, numbers, currencies, dates/times, etc that the Dart intl package does. That is in fact what the flutter_localizations package does for the Flutter SDK.

**Comment by HansMuller** [comment]
@marcglasberg per https://github.com/flutter/flutter/issues/41437#issuecomment-545579966

Using static extension methods for simple message translation is a nice idea and it does look pretty slick. Personally, I'm a bit leery of static extensions methods in libraries because, with heavy use, they can make code difficult to read. Glad to hear that it's working for your project!

**Comment by marcglasberg** [comment]
@HansMuller Instead of your Localization class having `ARB` as an intermediate format, it could instead have some simple map based format **in memory**. Then you could define an `Importer` interface, and anyone could then create importers to any formats. `ArbImporter` could be the first importer.

This would make it easy to create importers, so that people could choose their own formats, or even publish packages with importers.

Also that would allow for any other internalization libraries like my own to actually use these same importers. For example, in my lib I would call the same `ArbImporter` to create the in memory format, and then it's easy for me to convert from memory to my own format.

In other words, these importers would become reusable code for other internalization libraries.

**Comment by HansMuller** [comment]
@marcglasberg per https://github.com/flutter/flutter/issues/41437#issuecomment-546544973

> Instead of your Localization class having ARB as an intermediate format, it could instead have some simple map based format in memory. 

I agree, that would be much better.

**Comment by shihaohong** [comment]
Hi all, I've created the tool with @HansMuller's original proposal. All feedback and ideas are welcome. You can also comment on the PR https://github.com/flutter/flutter/pull/43934 if there are any questions or concerns.

## Tool Functionality

Starting with arb files in `lib/l10n`, the tool creates the following by default:
1. An `AppLocalizations` class containing:
- an `AppLocalizationsDelegate`
- a `supportedLocales` property that returns a list of all supported locales.
- a `localizationsDelegate` property that returns a list containing this delegate along with GlobalMaterialLocalizations.delegate, GlobalCupertinoLocalizations.delegate, and GlobalWidgetsLocalizations.delegate.
2. One Dart `message_<locale>.dart` file for each arb file, as well as a `messages_all.dart` file that performs the locale message lookup. The tool infers the locale of each arb file from the @@locale property and the arb file's name.
All feedback and ideas are welcome. You can also comment on the PR https://github.com/flutter/flutter/pull/43934 if there are any questions or concerns.
Once the tool is done running, all you would to do is modify the `MaterialApp.localizationsDelegates` and `MaterialApp.supportedLocales` properties.

## Tool Demo
1. Go to https://github.com/shihaohong/i18n_exploration and clone down the repo.
2. From the project root directory, run:
```
dart ${FLUTTER_DIR}/dev/tools/localization/gen_l10n.dart
```
3. Note that the above files have been generated in the `${PROJECT_REPO}/lib/l10n` folder.
4. The modifications to the app to accept localizations have already been added, but typically, you'd have to make the following changed to your app: 
```
import 'l10n/app_localizations.dart'; // add this line
// ...
class I18NDemoApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      localizationsDelegates: AppLocalizations.localizationsDelegates, // add this line
      supportedLocales: AppLocalizations.supportedLocales, // add this line
      home: Home(),
    );
  }
}
```

5. You should then be able to access localized messages depending on the locale of your simulator or physical device:
```
class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Text(AppLocalizations.of(context).helloWorld),
            Text(AppLocalizations.of(context).hello('International World')),
            Text(AppLocalizations.of(context).greeting('Greetings', 'Earthlings')),
            Text(AppLocalizations.of(context).helloWorlds(100)),
          ],
        ),
      ),
    );
  }
}
```

## Tool Options
There are a couple of modifiable options for the tool: 
1. `--arb-dir`: the path to your l10n directory. The tool defaults to `/lib/l10n/`.
2. `--template-arb-file`: the arb file to use as a template to generate these files. The tool defaults to `app_en.arb`.
3. `--output-localization-file`: The filename of the Dart file containing your localization and localizations delegate classes. The tool defaults to `app_localizations.dart`.
4. `--output-class`: The Dart class name to use for your localization class. The tool defaults to `AppLocalizations`.

**Comment by shihaohong** [comment]
I've added proposals [to a new GitHub issue](https://github.com/flutter/flutter/issues/47008) to how we could handle DateTime, Numbers and Currencies to the tool. 

Please take a look and provide any feedback, suggestions, or nice-to-haves you'd like to see if you have any :)

**Comment by goranluledzija** [comment]
We published [Flutter Intl](https://marketplace.visualstudio.com/items?itemName=localizely.flutter-intl) extension for VS Code that watches ARB files and generates localization code automatically.

So after adding string keys to ARB files, key getters are immediately available for autocomplete in Dart code.

Would like to hear your feedback, suggestions...

**Comment by shihaohong** [comment]
I just added an enhancement to the tool to [support a preferred locales option](https://github.com/flutter/flutter/pull/47845). By default, the supported locale list is sorted by alphabetical order. This option allows developer to set a prioritized list of locales to be placed at the beginning of the list over the rest of the sorted locales.

**Comment by mehmetf** [comment]
- Have you looked into how this proposal will work in google3?
- Does this proposal take into account that we will make embedder be more responsible about locale selection? https://github.com/flutter/flutter/issues/47087

**Comment by CalvinKern** [comment]
I looked over the PR for generating the AppLocalizations class, it doesn't seem like it supports locales with a scriptCode. We have a few different locales we use as a "custom" language pack.

i.e., intl_en_instukhe_GB.arb would have the locale:
`Locale.fromSubtags(languageCode: 'en', scriptCode: 'instukhe', countryCode: 'GB')`

We were able to get this by overriding our AppLocalizations class to check for these locales when `isSupported` is called, as well as providing these custom locales to the MaterialApp's supported locales.

Does this simplification of generating AppLocalizations remove the support for custom language handling?

**Comment by orestesgaolin** [comment]
Support for ARB is coming to more and more tools as this is the default for Flutter. Recently there was new open source tool announced called [Arbify](https://github.com/Arbify/Arbify). This is a self-hosted tool that can be run in Docker container, supports multiple projects, plural and genders.

**Comment by HansMuller** [comment]
@orestesgaolin - thanks for the pointers and for the reminder that it's time to acknowledge that the work proposed in this PR has been completed. The last i18n tooling change was https://github.com/flutter/flutter/pull/57510, and the completed system part of 1.19 beta and will land in the next stable release as well. 

There's a user guide here: https://flutter.dev/go/i18n-user-guide

**Comment by jbojcic1** [comment]
@HansMuller when can we expect the next stable release which will include this?

**Comment by HansMuller** [comment]
@csells - can you let @jbojcic1 et al. know when to expect the next stable release?

**Comment by csells** [comment]
@jbojcic1 we haven't announced the next stable release, but you should expect it in the next month or two.

**Comment by jsroest** [comment]
@shihaohong and @HansMuller 
Error: Could not resolve the package 'args' in 'package:args/args.dart'.

When following this guide https://flutter.dev/go/i18n-user-guide step-by-step (on the beta channel), I ran into the same problem as described in https://github.com/flutter/samples/issues/331

Running: flutter update-packages in ~/development/flutter/packages/flutter, solved it for me.
https://github.com/flutter/samples/issues/331#issuecomment-587891674

**Comment by kuhnroyal** [comment]
Running `flutter update-packages` in the SDK is not really a solution. Especially in CI when switching between Flutter versions a lot. This should be fetched when installing by default.

**Comment by shihaohong** [comment]
@kuhnroyal 
> 1. I need to manually run pub get in flutter/dev/tools for the generator to work - Will this automatically be installed in the future? Since the whole process starts by itself when I add the i10n.yaml file, I would expect it to be ready to use at that point.

I'm currently looking into fixing this. This is a result of the gen_l10n tool itself being a package, which is why it requires a `pub get` call since its dependencies have changed between flutter versions. I'm looking into seeing if I can pre-compile the dependencies with the tool itself so that it does not require re-caching and updating its dependencies.

> 2. What is the recommendation for the generated Dart files, should they be committed to VCS?

If you're going the synthetic package route (which we're trying to make the default way of doing things through https://github.com/flutter/flutter/pull/62395), then the goal is that generated Dart files will not need to be committed to VCS. If you'd still prefer to check in the generated code into VCS, you can turn off the synthetic package option with the ways mentioned in the PR depending on whether you're using an l10n.yaml file or running the tool directly via the command line.

**Comment by kuhnroyal** [comment]
@shihaohong Thanks for the clarification and your work on this!
I would love to not commit the generated files but in my initial tries I ran into problems running `flutter test` in CI. The files are not generated before running tests, that would currently leave me with the option to run the tool directly beforehand or committing the files.
Are the plans to make this tool available as generator so it can be tied into a `build_runner` call?

**Comment by shihaohong** [comment]
@kuhnroyal Actually, the current issue I'm trying to resolve is getting `flutter analyze` to figure out this situation. What's happening is that `flutter analyze` fails sometimes because the l10n code had not been generated beforehand. So, I'm trying to get `flutter analyze` to run `flutter pub get` as a step (which I'm trying to have call the synthetic package generation code) so that it does not fail for this reason. 

I had not considered that if would fail with `flutter test`. Could you share more about the kind of problems you're running into? Also, what does `build_runner` do with code generation that you would like to see? (I'm not as familiar with using `build_runner`).

**Comment by kuhnroyal** [comment]
@shihaohong From my understanding with the synthetic route (i10n.yaml) I would not commit the files to VCS. When I checkout a project I am usually able to run `flutter pub get` and `flutter test` which would compile the code and run the tests. But now the localization files are missing as they are only being generated during `flutter run/build` so the tests don't compile.

The question about `build_runner` is related to my test problem. Many projects use `build_runner` generators to generate code, mostly `*.g,dart` files for all kinds of libraries, built_value, moor, freezed and lots of other i18n libraries. In most cases the `*.g.dart` files are also not committed so that a dedicated build step is required before running tests or building the project (`flutter pub run build_runner build`) and many CI pipelines in a lot of projects already use this. There is already an established config file format `build.yaml` via https://pub.dev/packages/build_config. I was just wondering if there are plans to provide a generator for build_runner.

**Comment by shihaohong** [comment]
@kuhnroyal Great, thanks for the feedback/explanation! I'll take a look into how we can address that in the next few days. I might create a separate issue to track it as well since this one is already closed and will loop you in once I do that.

**Comment by shihaohong** [comment]
@kuhnroyal My synthetic package PR (https://github.com/flutter/flutter/pull/62395) incorporates synthetic package generation into `flutter test` and `flutter analyze` invocations when [`flutter pub get` is required](https://github.com/flutter/flutter/pull/62395/files#diff-5b0138e4d7db9e5a88bcbe1ab7f93dc0R96), so that should address the issue you mentioned with running tests and compiling code.

As for providing a generator for `build_runner`, we're currently not planning on incorporating synthetic package generation into `build_runner`, but this shouldn't be a problem as long as there isn't a problem with the synthetic package generation code. Please feel free to file a new issue highlighting any concerns or problems you face with the `gen_l10n` tool, I'd be happy to look into and try to address them.

**Comment by Sominemo** [comment]
> https://flutter.dev/go/i18n-user-guide

I noticed the `output-dir` parameter from the docs doesn't exist in code

**Comment by shihaohong** [comment]
@Sominemo  Thanks for the heads up! I need to make some updates to the user guide, some of the instructions have changed with some updates to `gen_l10n`.

**Comment by kuhnroyal** [comment]
@shihaohong Did this make it into the new beta?

**Comment by shihaohong** [comment]
Yes, it looks like the [synthetic packages commit made it into beta](https://github.com/flutter/flutter/commits/beta?after=8b3760638a189741cd9ca881aa2dd237c1df1be5+34&branch=beta).

**Comment by kuhnroyal** [comment]
@shihaohong The synthetic package seems to break `build_runner` as it finds the `.dart_tool/flutter_gen` folder and expects a `pubspec.yaml` there.

```
Bad state: Unable to generate package graph, no `/Users/foo/workspace/test/.dart_tool/flutter_gen/pubspec.yaml` found.
#0      _pubspecForPath (package:build_runner_core/src/package_graph/package_graph.dart:235:5)
#1      _parsePackageDependencies (package:build_runner_core/src/package_graph/package_graph.dart:208:21)
#2      PackageGraph.forPath (package:build_runner_core/src/package_graph/package_graph.dart:97:33)
<asynchronous suspension>
#3      PackageGraph.forThisPackage (package:build_runner_core/src/package_graph/package_graph.dart:110:20)
#4      main (file:///Users/foo/.../build_runner-1.10.2/bin/build_runner.dart:27:49)
#5      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
#6      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
pub finished with exit code 255

```

**Comment by shihaohong** [comment]
@kuhnroyal The first issue you mentioned was also mentioned/filed in https://github.com/dart-lang/build/issues/2835.

For the next two, could you create separate issues for them and tag me? It's easier for us to keep track of new issues with the tool that way since these problems can get lost in this issue's comment thread.
